File: test/watcher_test.dart
import 'dart:async';
import 'dart:io';
import 'package:test/test.dart';
import 'package:sidelook/src/watcher.dart';
import 'package:path/path.dart' as p;

/// Helper para indicar que um Future não será awaited intencionalmente
void unawaited(Future<void> future) {}

void main() {
  group('isImageFile', () {
    test('aceita jpg', () => expect(isImageFile('foto.jpg'), isTrue));
    test('aceita jpeg', () => expect(isImageFile('foto.jpeg'), isTrue));
    test('aceita png',
        () => expect(isImageFile('foto.PNG'), isTrue)); // case insensitive
    test('aceita gif', () => expect(isImageFile('animacao.gif'), isTrue));
    test('aceita webp', () => expect(isImageFile('imagem.webp'), isTrue));
    test('aceita svg', () => expect(isImageFile('vetor.svg'), isTrue));
    test('aceita bmp', () => expect(isImageFile('bitmap.bmp'), isTrue));
    test('aceita tiff', () => expect(isImageFile('scan.tiff'), isTrue));
    test('aceita tif', () => expect(isImageFile('scan.tif'), isTrue));

    test('rejeita txt', () => expect(isImageFile('arquivo.txt'), isFalse));
    test('rejeita pdf', () => expect(isImageFile('doc.pdf'), isFalse));
    test('rejeita sem extensão', () => expect(isImageFile('arquivo'), isFalse));
  });

  group('ImageWatcher', () {
    late Directory tempDir;

    setUp(() async {
      tempDir = await Directory.systemTemp.createTemp('sidelook_test_');
    });

    tearDown(() async {
      await tempDir.delete(recursive: true);
    });

    test('validate retorna true para diretório válido', () async {
      final watcher = ImageWatcher(tempDir.path);
      expect(await watcher.validate(), isTrue);
    });

    test('validate retorna false para diretório inexistente', () async {
      final watcher = ImageWatcher('/caminho/que/nao/existe');
      expect(await watcher.validate(), isFalse);
    });

    test('scanExisting encontra imagem mais recente', () async {
      // Criar algumas imagens com diferentes timestamps
      final img1 = File(p.join(tempDir.path, 'img1.png'));
      await img1.writeAsBytes([0x89, 0x50, 0x4E, 0x47]); // PNG magic bytes
      await Future<void>.delayed(Duration(seconds: 1));

      final img2 = File(p.join(tempDir.path, 'img2.jpg'));
      await img2.writeAsBytes([0xFF, 0xD8, 0xFF]); // JPEG magic bytes

      final watcher = ImageWatcher(tempDir.path);
      final result = await watcher.scanExisting();

      expect(result.count, equals(2));
      expect(result.mostRecent?.path, equals(img2.path));
    });

    test('scanExisting retorna null se não há imagens', () async {
      // Criar arquivo não-imagem
      final txt = File(p.join(tempDir.path, 'readme.txt'));
      await txt.writeAsString('Hello');

      final watcher = ImageWatcher(tempDir.path);
      final result = await watcher.scanExisting();

      expect(result.count, equals(0));
      expect(result.mostRecent, isNull);
    });

    test('detecta nova imagem', () async {
      final watcher = ImageWatcher(tempDir.path);
      await watcher.start();

      // Criar listener antes de adicionar arquivo
      final completer = Completer<File>();
      unawaited(watcher.onNewImage.first.then(completer.complete));

      // Dar tempo para watcher inicializar
      await Future<void>.delayed(Duration(milliseconds: 100));

      // Criar nova imagem
      final newImg = File(p.join(tempDir.path, 'nova.png'));
      await newImg.writeAsBytes([0x89, 0x50, 0x4E, 0x47]);

      // Aguardar detecção (com timeout)
      final detected = await completer.future.timeout(
        Duration(seconds: 5),
        onTimeout: () => throw TimeoutException('Imagem não detectada'),
      );

      expect(detected.path, equals(newImg.path));

      await watcher.stop();
    }, timeout: Timeout(Duration(seconds: 10)));

    test('detecta remoção da imagem atual e atualiza para próxima', () async {
      final tempDir = await Directory.systemTemp.createTemp('sidelook_test_');
      addTearDown(() => tempDir.delete(recursive: true));

      // Criar duas imagens com delay para garantir ordem de modificação
      final img1 = File(p.join(tempDir.path, 'primeira.jpg'));
      await img1.writeAsBytes([0xFF, 0xD8, 0xFF, 0xE0]);
      await Future<void>.delayed(Duration(seconds: 1));

      final img2 = File(p.join(tempDir.path, 'segunda.jpg'));
      await img2.writeAsBytes([0xFF, 0xD8, 0xFF, 0xE0]);

      final watcher = ImageWatcher(tempDir.path);
      final result = await watcher.scanExisting();

      // img2 deve ser a mais recente
      expect(result.mostRecent?.path, equals(img2.path));
      expect(watcher.currentImage?.path, equals(img2.path));

      await watcher.start();

      // Configurar listener para capturar a próxima imagem
      final completer = Completer<File>();
      unawaited(watcher.onNewImage.first.then(completer.complete));

      // Dar tempo para watcher inicializar
      await Future<void>.delayed(Duration(milliseconds: 100));

      // Remover a imagem atual (img2)
      await img2.delete();

      // Aguardar que o watcher detecte e notifique a próxima imagem
      final nextImage = await completer.future.timeout(
        Duration(seconds: 5),
        onTimeout: () => throw TimeoutException('Próxima imagem não detectada'),
      );

      // Deve notificar img1 como a nova imagem mais recente
      expect(nextImage.path, equals(img1.path));
      expect(watcher.currentImage?.path, equals(img1.path));

      await watcher.stop();
    }, timeout: Timeout(Duration(seconds: 10)));

    test('detecta remoção quando não há mais imagens', () async {
      final tempDir = await Directory.systemTemp.createTemp('sidelook_test_');
      addTearDown(() => tempDir.delete(recursive: true));

      // Criar apenas uma imagem
      final img = File(p.join(tempDir.path, 'unica.jpg'));
      await img.writeAsBytes([0xFF, 0xD8, 0xFF, 0xE0]);

      final watcher = ImageWatcher(tempDir.path);
      await watcher.scanExisting();
      expect(watcher.currentImage?.path, equals(img.path));

      await watcher.start();
      await Future<void>.delayed(Duration(milliseconds: 100));

      // Remover a única imagem
      await img.delete();

      // Aguardar processamento
      await Future<void>.delayed(Duration(seconds: 2));

      // currentImage deve ser null
      expect(watcher.currentImage, isNull);

      await watcher.stop();
    }, timeout: Timeout(Duration(seconds: 10)));

    test('ignora remoção de imagem que não é a atual', () async {
      final tempDir = await Directory.systemTemp.createTemp('sidelook_test_');
      addTearDown(() => tempDir.delete(recursive: true));

      // Criar duas imagens
      final img1 = File(p.join(tempDir.path, 'primeira.jpg'));
      await img1.writeAsBytes([0xFF, 0xD8, 0xFF, 0xE0]);
      await Future<void>.delayed(Duration(seconds: 1));

      final img2 = File(p.join(tempDir.path, 'segunda.jpg'));
      await img2.writeAsBytes([0xFF, 0xD8, 0xFF, 0xE0]);

      final watcher = ImageWatcher(tempDir.path);
      await watcher.scanExisting();

      // img2 é a atual
      expect(watcher.currentImage?.path, equals(img2.path));

      await watcher.start();
      await Future<void>.delayed(Duration(milliseconds: 100));

      // Remover img1 (não é a atual)
      await img1.delete();

      // Aguardar
      await Future<void>.delayed(Duration(seconds: 1));

      // currentImage ainda deve ser img2
      expect(watcher.currentImage?.path, equals(img2.path));

      await watcher.stop();
    }, timeout: Timeout(Duration(seconds: 10)));
  });
}

File: sidelook/internal/watcher/watcher_test.go
// internal/watcher/watcher_test.go
package watcher

import (
	"os"
	"path/filepath"
	"testing"
	"time"
)

func TestIsImageFile(t *testing.T) {
	tests := []struct {
		path     string
		expected bool
	}{
		{"foto.jpg", true},
		{"foto.jpeg", true},
		{"foto.PNG", true}, // case insensitive
		{"animacao.gif", true},
		{"imagem.webp", true},
		{"vetor.svg", true},
		{"bitmap.bmp", true},
		{"scan.tiff", true},
		{"scan.tif", true},
		{"arquivo.txt", false},
		{"doc.pdf", false},
		{"arquivo", false},
	}

	for _, tt := range tests {
		t.Run(tt.path, func(t *testing.T) {
			result := IsImageFile(tt.path)
			if result != tt.expected {
				t.Errorf("IsImageFile(%q) = %v, want %v", tt.path, result, tt.expected)
			}
		})
	}
}

func TestImageWatcher_New(t *testing.T) {
	// Diretório válido
	tmpDir, err := os.MkdirTemp("", "sidelook_test_*")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(tmpDir)

	w, err := New(tmpDir)
	if err != nil {
		t.Errorf("New() error = %v, want nil", err)
	}
	if w != nil {
		w.Stop()
	}

	// Diretório inválido
	_, err = New("/caminho/que/nao/existe")
	if err == nil {
		t.Error("New() com diretório inválido deveria retornar erro")
	}
}

func TestImageWatcher_ScanExisting(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "sidelook_test_*")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(tmpDir)

	// Criar algumas imagens
	img1 := filepath.Join(tmpDir, "img1.png")
	if err := os.WriteFile(img1, []byte{0x89, 0x50, 0x4E, 0x47}, 0644); err != nil {
		t.Fatal(err)
	}

	time.Sleep(100 * time.Millisecond)

	img2 := filepath.Join(tmpDir, "img2.jpg")
	if err := os.WriteFile(img2, []byte{0xFF, 0xD8, 0xFF}, 0644); err != nil {
		t.Fatal(err)
	}

	// Criar arquivo não-imagem
	txt := filepath.Join(tmpDir, "readme.txt")
	if err := os.WriteFile(txt, []byte("hello"), 0644); err != nil {
		t.Fatal(err)
	}

	w, err := New(tmpDir)
	if err != nil {
		t.Fatal(err)
	}
	defer w.Stop()

	count, mostRecent, err := w.ScanExisting()
	if err != nil {
		t.Errorf("ScanExisting() error = %v", err)
	}

	if count != 2 {
		t.Errorf("ScanExisting() count = %d, want 2", count)
	}

	if mostRecent == nil {
		t.Fatal("ScanExisting() mostRecent = nil, want non-nil")
	}

	if mostRecent.Path != img2 {
		t.Errorf("ScanExisting() mostRecent.Path = %q, want %q", mostRecent.Path, img2)
	}
}

func TestImageWatcher_DetectNewImage(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "sidelook_test_*")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(tmpDir)

	w, err := New(tmpDir)
	if err != nil {
		t.Fatal(err)
	}
	defer w.Stop()

	detected := make(chan string, 1)
	w.OnNewImage = func(path string) {
		detected <- path
	}

	if err := w.Start(); err != nil {
		t.Fatal(err)
	}

	// Dar tempo para watcher inicializar
	time.Sleep(100 * time.Millisecond)

	// Criar nova imagem
	newImg := filepath.Join(tmpDir, "nova.png")
	if err := os.WriteFile(newImg, []byte{0x89, 0x50, 0x4E, 0x47}, 0644); err != nil {
		t.Fatal(err)
	}

	// Aguardar detecção
	select {
	case path := <-detected:
		if path != "nova.png" {
			t.Errorf("OnNewImage path = %q, want %q", path, "nova.png")
		}
	case <-time.After(5 * time.Second):
		t.Error("Timeout aguardando detecção de nova imagem")
	}
}
